File: Swift/ExampleTests/common/ShoppingCartTyped.swift
public struct ShoppingCartTyped: Equatable, CustomStringConvertible {
    public let items: [OrderItemTyped]
    public let shipping: String
    public let discount: String
    public let totalPrice: String
    public let shippingAddress: AddressTyped
    public let billingAddress: AddressTyped

    public init(items: [OrderItemTyped], shipping: String, discount: String, totalPrice: String, shippingAddress: AddressTyped, billingAddress: AddressTyped) {
        self.items = items
        self.shipping = shipping
        self.discount = discount
        self.totalPrice = totalPrice
        self.shippingAddress = shippingAddress
        self.billingAddress = billingAddress
    }

    public init(from s: ShoppingCartString) {
        self.items = []
        self.shipping = s.shipping
        self.discount = s.discount
        self.totalPrice = s.totalPrice
        self.shippingAddress = AddressTyped(from: s.shippingAddress)
        self.billingAddress = AddressTyped(from: s.billingAddress)
    }

    public func toStringStruct() -> ShoppingCartString {
        return ShoppingCartString(
            items: "",
            shipping: String(describing: shipping),
            discount: String(describing: discount),
            totalPrice: String(describing: totalPrice),
            shippingAddress: shippingAddress.toStringStruct(),
            billingAddress: billingAddress.toStringStruct()
        )
    }

    public static func toStringList(_ list: [ShoppingCartTyped]) -> [ShoppingCartString] {
        return list.map { $0.toStringStruct() }
    }

    public static func fromStringList(_ list: [ShoppingCartString]) -> [ShoppingCartTyped] {
        return list.map { ShoppingCartTyped(from: $0) }
    }

    public func toJSONValue() -> [String: Any] {
        return [
            "items": items.map { $0.toJSONValue() },
            "shipping": shipping,
            "discount": discount,
            "totalPrice": totalPrice,
            "shippingAddress": shippingAddress.toJSONValue(),
            "billingAddress": billingAddress.toJSONValue(),
        ]
    }

    public func toJSON() throws -> String {
        return try Json.write(toJSONValue())
    }

    public init(fromJSONValue m: [String: Any]) throws {
        self.items = try Json.asArray(Json.require(m, "items"), "items").map { try OrderItemTyped(fromJSONValue: Json.asObject($0, "items")) }
        self.shipping = try Json.asString(Json.require(m, "shipping"), "shipping")
        self.discount = try Json.asString(Json.require(m, "discount"), "discount")
        self.totalPrice = try Json.asString(Json.require(m, "totalPrice"), "totalPrice")
        self.shippingAddress = try AddressTyped(fromJSONValue: Json.asObject(Json.require(m, "shippingAddress"), "shippingAddress"))
        self.billingAddress = try AddressTyped(fromJSONValue: Json.asObject(Json.require(m, "billingAddress"), "billingAddress"))
    }

    public init(fromJSON text: String) throws {
        try self.init(fromJSONValue: Json.parseObject(text))
    }

    public static func toJSONList(_ list: [ShoppingCartTyped]) throws -> String {
        return try Json.write(list.map { $0.toJSONValue() })
    }

    public static func fromJSONList(_ text: String) throws -> [ShoppingCartTyped] {
        return try Json.parseArray(text).map {
            try ShoppingCartTyped(fromJSONValue: Json.asObject($0, "ShoppingCartTyped"))
        }
    }

    public var description: String {
        return "Items=\(items), Shipping=\(shipping), Discount=\(discount), TotalPrice=\(totalPrice), ShippingAddress=\(shippingAddress), BillingAddress=\(billingAddress)"
    }
}

File: Rust/ExampleTests/common/shoppingcart_typed.rs
#![allow(dead_code, unused_imports, unused_variables)]

use super::json;
use super::shoppingcart_string::ShoppingCartString;
use super::orderitem_typed::OrderItemTyped;
use super::address_typed::AddressTyped;

#[derive(Debug, Clone, Default, PartialEq)]
pub struct ShoppingCartTyped {
    pub items: Vec<OrderItemTyped>,
    pub shipping: String,
    pub discount: String,
    pub totalprice: String,
    pub shippingaddress: AddressTyped,
    pub billingaddress: AddressTyped,
}

impl ShoppingCartTyped {
    pub fn from_str_struct(s: &ShoppingCartString) -> Self {
        Self {
            items: Vec::new(),
            shipping: s.shipping.clone(),
            discount: s.discount.clone(),
            totalprice: s.totalprice.clone(),
            shippingaddress: AddressTyped::from_str_struct(&s.shippingaddress),
            billingaddress: AddressTyped::from_str_struct(&s.billingaddress),
        }
    }

    pub fn to_str_struct(&self) -> ShoppingCartString {
        ShoppingCartString {
            items: String::new(),
            shipping: self.shipping.clone(),
            discount: self.discount.clone(),
            totalprice: self.totalprice.clone(),
            shippingaddress: self.shippingaddress.to_str_struct(),
            billingaddress: self.billingaddress.to_str_struct(),
        }
    }

    pub fn to_string_list(list: &[ShoppingCartTyped]) -> Vec<ShoppingCartString> {
        list.iter().map(|t| t.to_str_struct()).collect()
    }

    pub fn from_string_list(list: &[ShoppingCartString]) -> Vec<ShoppingCartTyped> {
        list.iter().map(ShoppingCartTyped::from_str_struct).collect()
    }

    pub fn to_json_value(&self) -> json::Value {
        json::Value::Object(vec![
            ("Items".to_string(), json::Value::Array(self.items.iter().map(|e| e.to_json_value()).collect())),
            ("Shipping".to_string(), json::Value::Str(self.shipping.clone())),
            ("Discount".to_string(), json::Value::Str(self.discount.clone())),
            ("TotalPrice".to_string(), json::Value::Str(self.totalprice.clone())),
            ("ShippingAddress".to_string(), self.shippingaddress.to_json_value()),
            ("BillingAddress".to_string(), self.billingaddress.to_json_value()),
        ])
    }

    pub fn to_json(&self) -> String {
        json::write(&self.to_json_value())
    }

    pub fn from_json_value(v: &json::Value) -> json::JsonResult<Self> {
        Ok(Self {
            items: json::as_array(json::require(v, "Items")?, "Items")?.iter().map(|e| OrderItemTyped::from_json_value(e)).collect::<json::JsonResult<Vec<_>>>()?,
            shipping: json::as_string(json::require(v, "Shipping")?, "Shipping")?,
            discount: json::as_string(json::require(v, "Discount")?, "Discount")?,
            totalprice: json::as_string(json::require(v, "TotalPrice")?, "TotalPrice")?,
            shippingaddress: AddressTyped::from_json_value(json::require(v, "ShippingAddress")?)?,
            billingaddress: AddressTyped::from_json_value(json::require(v, "BillingAddress")?)?,
        })
    }

    pub fn from_json(text: &str) -> json::JsonResult<Self> {
        Self::from_json_value(&json::parse(text)?)
    }

    pub fn to_json_list(list: &[ShoppingCartTyped]) -> String {
        json::write(&json::Value::Array(
            list.iter().map(|item| item.to_json_value()).collect(),
        ))
    }

    pub fn from_json_list(text: &str) -> json::JsonResult<Vec<ShoppingCartTyped>> {
        let root = json::parse(text)?;
        let items = json::as_array(&root, "ShoppingCartTyped")?;
        let mut result = Vec::with_capacity(items.len());
        for e in items {
            result.push(Self::from_json_value(e)?);
        }
        Ok(result)
    }
}
